plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+                    raise ValueError(
+                        f"Input arrays must be finite: {rq} contains "
+                        "either NaN or infinite values."
+                    )
+
+                # Check that the max values on the edges of the arrays are
+                # small relative to the maximum values on that grid
+                #
+                # Array must be dimensionless to re-assemble it into an array
+                # of max values like this
+                arr = np.abs(grid[rq]).value
+                edge_max = np.max(
+                    np.array(
+                        [
+                            np.max(a)
+                            for a in (
+                                arr[0, :, :],
+                                arr[-1, :, :],
+                                arr[:, 0, :],
+                                arr[:, -1, :],
+                                arr[:, :, 0],
+                                arr[:, :, -1],
+                            )
+                        ]
+                    )
+                )
+
+                if edge_max > 1e-3 * np.max(arr):
+                    unit = grid.recognized_quantities[rq].unit
+                    warnings.warn(
+                        "Fields should go to zero at edges of grid to avoid "
+                        f"non-physical effects, but a value of {edge_max:.2E} {unit} was "
+                        f"found on the edge of the {rq} array. Consider applying a "
+                        "envelope function to force the fields at the edge to go to "
+                        "zero.",
+                        RuntimeWarning,
+                    )
+
+        # Validate inputs to the run function
+        # Sets is_synchronized_time property
+        self._validate_constructor_inputs(
+            termination_condition, save_routine, dt, dt_range, field_weighting
+        )
+
+        self.dt = dt.to(u.s).value if dt is not None else None
+
+        dt_range = [0, np.inf] * u.s if dt_range is None else dt_range
+        self.dt_range = dt_range.to(u.s).value
+
+        # Update the `tracker` attribute so that the stop condition & save routine can be used
+        termination_condition.tracker = self
+
+        if save_routine is not None:
+            save_routine.tracker = self
+
+        self.termination_condition = termination_condition
+        self.save_routine = save_routine
+
+    @property
+    def num_grids(self):
+        """The number of grids specified at instantiation."""
+        return len(self.grids)
+
+    def setup_adaptive_time_step(
+        self,
+        time_steps_per_gyroperiod: Optional[int] = 12,
+        Courant_parameter: Optional[float] = 0.5,
+    ):
+        """Set parameters for the adaptive time step candidates.
+
+        Parameters
+        ----------
+        time_steps_per_gyroperiod : int, optional
+            The number of subdivisions of the particle's gyration. The default is twelve.
```

Comment:
The minimum ratio of the particle gyroperiod to the timestep. Higher numbers correspond to higher temporal resolution.